src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
+                        this.Reset(ref resetPoint2);
+                        this.Release(ref resetPoint2);
+                    }
+                }
+
+                if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken && (type != null || !looksLikeCast()))
+                {
+                    // It is possible this is a parenthesized (constant) expression.
+                    // We normalize later.
+                    ParseSubpatternList(out var openParenToken, out var subPatterns, out var closeParenToken, SyntaxKind.OpenParenToken, SyntaxKind.CloseParenToken);
+                    if (this.CurrentToken.Kind == SyntaxKind.EqualsGreaterThanToken)
+                    {
+                        // Testers do the darndest things, in this case putting a lambda expression where a pattern is expected.
                         this.Reset(ref resetPoint);
-                        node = this.ParseSubExpression(Precedence.Expression);
+                        return this.ParseSubExpression(Precedence.Expression);
+                    }
+
+                    parsePropertySubpattern();
+                    parseDesignation();
+
+                    if (type == null &&
+                        propertySubpattern == null &&
+                        designation == null &&
+                        subPatterns.Count == 1 &&
+                        subPatterns[0].NameColon == null &&
+                        subPatterns[0].Pattern is ConstantPatternSyntax cp)
+                    {
+                        // There is an ambiguity between a deconstruction pattern `(` pattern `)`
+                        // and a constant expression pattern than happens to be parenthesized.
+                        // We normalize to the parenthesized expression, and semantic analysis
+                        // will notice the parens and treat it whichever way is semantically sensible,
+                        // giving priority to its treatment as a constant expression.
+                        return _syntaxFactory.ParenthesizedExpression(openParenToken, cp.Expression, closeParenToken);
                     }
+
+                    var node = _syntaxFactory.DeconstructionPattern(type, openParenToken, subPatterns, closeParenToken, propertySubpattern, designation);
+                    return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
                 }
-                finally
+
+                if (parsePropertySubpattern())
                 {
-                    this.Release(ref resetPoint);
+                    parseDesignation();
+                    var node = _syntaxFactory.PropertyPattern(type, propertySubpattern, designation);
+                    return this.CheckFeatureAvailability(node, MessageID.IDS_FeatureRecursivePatterns);
+                }
+
+                if (type != null && parseDesignation())
+                {
+                    return _syntaxFactory.DeclarationPattern(type, designation);
```

Comment:
>return _syntaxFactory.DeclarationPattern(type, designation); [](start = 20, length = 60)

Do we need to check feature availability for declaration pattern?

Reply:
No. We do so in each construct which has a pattern as a component.

---
In reply to: [151814906](https://github.com/dotnet/roslyn/pull/23208#discussion_r151814906) [](ancestors = 151814906)